### PR TITLE
Fix: liquidation & collateral edge cases

### PR DESCRIPTION
**Pull request by sander2**

### CODE REVIEW

File: crates/vault-registry/src/lib.rs [hunk 1]
```diff
@@ -308,38 +326,132 @@ impl<T: Config> Module<T> {
...
         Self::deposit_event(Event::<T>::IncreaseToBeIssuedTokens(vault.id(), tokens));
-        let btc_address = vault.new_deposit_address(secure_id)?;
+        Ok(())
+    }
+
+    /// Registers a btc address
+    ///
+    /// # Arguments
+    /// * `issue_id` - secure id for generating deposit address
+    pub fn _register_address(
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:428]: Naming is confusing because the similarly named dispatchable function above, maybe we could rename to `register_deposit_address` unless you have another idea?

File: crates/vault-registry/src/lib.rs [hunk 2]
```diff
@@ -390,6 +502,14 @@ impl<T: Config> Module<T> {
         Self::get_rich_liquidation_vault().force_decrease_issued(tokens)
     }
 
+    /// Increases the amount of tokens issued for the liquidation vault
+    ///
+    /// # Arguments
+    /// * `tokens` - the amount of tokens to be unreserved
+    pub fn liquidation_vault_force_increase_issued_tokens(tokens: PolkaBTC<T>) -> DispatchResult {
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:509]:
```suggestion
    pub fn increase_liquidation_vault_issued_tokens(tokens: PolkaBTC<T>) -> DispatchResult {
```

File: crates/issue/src/ext.rs [hunk 3]
```diff
@@ -63,19 +77,31 @@ pub(crate) mod vault_registry {
...
-    }
-
     pub fn ensure_not_banned<T: vault_registry::Config>(
         vault: &T::AccountId,
         height: T::BlockNumber,
     ) -> DispatchResult {
         <vault_registry::Module<T>>::_ensure_not_banned(vault, height)
     }
+
+    pub fn liquidation_vault_force_decrease_to_be_issued_tokens<T: vault_registry::Config>(
```
Comment (gregdhill) [inline crates/issue/src/ext.rs:87]:
```suggestion
    pub fn decrease_liquidation_vault_to_be_issued_tokens<T: vault_registry::Config>(
```

File: crates/issue/src/ext.rs [hunk 4]
```diff
@@ -63,19 +77,31 @@ pub(crate) mod vault_registry {
...
+        amount: PolkaBTC<T>,
+    ) -> DispatchResult {
+        <vault_registry::Module<T>>::liquidation_vault_force_decrease_to_be_issued_tokens(amount)
+    }
+
+    pub fn liquidation_vault_force_increase_issued_tokens<T: vault_registry::Config>(
```
Comment (gregdhill) [inline crates/issue/src/ext.rs:93]:
```suggestion
    pub fn increase_liquidation_vault_issued_tokens<T: vault_registry::Config>(
```

File: crates/vault-registry/src/lib.rs [hunk 5]
```diff
@@ -634,6 +751,16 @@ impl<T: Config> Module<T> {
         Ok(())
     }
 
+    /// Decreases to-be-issued tokens on vault (or liquidation vault, if it has been liquidated)
+    pub fn cancel_issue_tokens(vault_id: &T::AccountId, tokens: PolkaBTC<T>) -> DispatchResult {
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:755]:
```suggestion
    pub fn decrease_to_be_issued_tokens(vault_id: &T::AccountId, tokens: PolkaBTC<T>) -> DispatchResult {
```

File: crates/vault-registry/src/types.rs [hunk 6]
```diff
@@ -25,6 +25,50 @@ pub enum Version {
     V1,
 }
 
+#[derive(Debug, PartialEq)]
+pub enum CurrencyType<T: frame_system::Config> {
```
Comment (gregdhill) [inline crates/vault-registry/src/types.rs:29]: Why do we need this?

File: crates/replace/src/lib.rs [hunk 7]
```diff
@@ -348,7 +354,7 @@ impl<T: Config> Module<T> {
         // Retrieve the ReplaceRequest as per the replaceId parameter from Vaults in the VaultRegistry
         let replace = Self::get_open_replace_request(&request_id)?;
 
-        // Check that caller of the function is indeed the to-be-replaced Vault as specified in the ReplaceRequest. Return ERR_UNAUTHORIZED error if this check fails.
+        // Check that caller of the function is int is not for the entire BTC holdings, check that the deed the to-be-replaced Vault as specified in the ReplaceRequest. Return ERR_UNAUTHORIZED error if this check fails.
```
Comment (gregdhill) [inline crates/replace/src/lib.rs:357]: Something wrong with this comment

File: crates/replace/src/lib.rs [hunk 8]
```diff
@@ -605,22 +610,46 @@ impl<T: Config> Module<T> {
...
+            let new_collateral = ext::vault_registry::get_backing_collateral::<T>(&new_vault_id)?
+                .checked_sub(&replace.collateral)
+                .ok_or(Error::<T>::ArithmeticUnderflow)?;
+
+            let required_collateral =
+                ext::vault_registry::get_required_collateral_for_vault::<T>(&new_vault_id)?;
+
+            if new_collateral >= required_collateral {
+                // Release the newVault's collateral associated with this ReplaceRequests
+                ext::vault_registry::_withdraw_collateral::<T>(&new_vault_id, replace.collateral)?;
```
Comment (gregdhill) [inline crates/replace/src/lib.rs:650]: This might be weird behavior if the new_vault specified additional collateral for the replace.

File: crates/vault-registry/src/lib.rs [hunk 9]
```diff
@@ -766,10 +989,28 @@ impl<T: Config> Module<T> {
...
+
+        // the current locked backing collateral by the vault
+        let collateral = Self::get_backing_collateral(vault_id)?;
+
+        // the currently issued tokens in PolkaBTC
+        let tokens = vault
+            .data
+            .issued_tokens
+            .checked_sub(&vault.data.to_be_redeemed_tokens)
+            .ok_or(Error::<T>::ArithmeticOverflow)?;
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:1007]:
```suggestion
            .ok_or(Error::<T>::ArithmeticUnderflow)?;
```

Comment (gregdhill) [reply crates/vault-registry/src/lib.rs:1007]: Do we want to saturate this?

File: crates/replace/src/lib.rs (hunk 8 above)
Comment (sander2) [reply crates/replace/src/lib.rs:650]: To me it makes sense - the additional collateral in accept_replace is specifically to back the PolkaBTC it will receive from the old_vault. If the replace is cancelled, it should therefore be withdrawn. If it wants to increase its collateral for general purpose use, it should manually call `lock_additional_collateral` imo

File: crates/vault-registry/src/lib.rs (hunk 9 above)
Comment (sander2) [reply crates/vault-registry/src/lib.rs:1007]: No I don't think so, since it should always be true that `to_be_redeemed <= issued`

File: parachain/runtime/src/lib.rs [hunk 10]
```diff
@@ -382,7 +382,7 @@ impl parachain_tokens::Config for Runtime {
 }
 
 parameter_types! {
-    pub const ExistentialDeposit: u128 = 500;
+    pub const ExistentialDeposit: u128 = 1;
```
Comment (gregdhill) [inline parachain/runtime/src/lib.rs:385]: Not `0`?

File: crates/vault-registry/src/types.rs [hunk 11]
```diff
@@ -113,20 +163,22 @@ pub struct SystemVault<AccountId, PolkaBTC> {
     pub to_be_redeemed_tokens: PolkaBTC,
 }
 
-impl<AccountId, BlockNumber, PolkaBTC: HasCompact + Default>
-    Vault<AccountId, BlockNumber, PolkaBTC>
+impl<AccountId, BlockNumber, PolkaBTC: HasCompact + Default, Dot: HasCompact + Default>
```
Comment (gregdhill) [inline crates/vault-registry/src/types.rs:166]: Very minor: capitalize `DOT` for consistency, will make find/replace easier when we generalize.

**Comment by gregdhill** [comment]
For the collateralization rate, can you quickly check that the custom RPC calls will return based on `backing_collateral` now?

File: crates/vault-registry/src/lib.rs [hunk 12]
```diff
@@ -676,6 +824,68 @@ impl<T: Config> Module<T> {
...
+    pub fn cancel_replace_tokens(
+        old_vault_id: &T::AccountId,
+        new_vault_id: &T::AccountId,
+        tokens: PolkaBTC<T>,
+    ) -> DispatchResult {
+        Self::check_parachain_not_shutdown_and_not_errors(
+            [
+                ErrorCode::InvalidBTCRelay,
+                ErrorCode::OracleOffline,
+                ErrorCode::Liquidation,
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:845]: Maybe we can do this in another PR, but we should remove this ErrorCode since it's no longer set.

File: parachain/runtime/src/lib.rs (hunk 10 above)
Comment (sander2) [reply parachain/runtime/src/lib.rs:385]: In all of the mocks it was set to `1`, which I guess makes sense - if a balance becomes less than one, it's ok if the account gets removed

File: crates/vault-registry/src/lib.rs (hunk 12 above)
Comment (sander2) [reply crates/vault-registry/src/lib.rs:845]: made a ticket for this

**Comment by sander2** [comment]
> For the collateralization rate, can you quickly check that the custom RPC calls will return based on `backing_collateral` now?

Yea I checked, they do.

**Review by gregdhill — APPROVED** [review]
